pkg/extensions/tbtc/liquidation_recovery.go

```diff
@@ -0,0 +1,282 @@
+			return "", fmt.Errorf(
+				"error deriving external chain path /0 from extended key: [%s]",
+				err,
+			)
+		}
+	}
+
+	requestedPublicKey, err := externalChain.Child(uint32(addressIndex))
+	if err != nil {
+		return "", fmt.Errorf(
+			"error deriving requested address index /0/%v from extended key: [%s]",
+			addressIndex,
+			err,
+		)
+	}
+
+	// Now to decide how we want to serialize the address...
+	var chainParams *chaincfg.Params
+	switch publicKeyDescriptor {
+	case "xpub", "ypub", "zpub":
+		chainParams = &chaincfg.MainNetParams
+	case "tpub", "upub", "vpub":
+		chainParams = &chaincfg.TestNet3Params
+	}
+
+	requestedAddress, err := requestedPublicKey.Address(chainParams)
+
+	var finalAddress btcutil.Address = requestedAddress
+	switch publicKeyDescriptor {
+	case "xpub", "tpub":
+		// Noop, the address is already correct
+	case "ypub", "upub":
+		// p2wpkh-in-p2sh, constructed as per https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#p2wpkh-nested-in-bip16-p2sh .
+		scriptSig := append([]byte{0x00, 0x14}, requestedAddress.Hash160()[:]...)
+		finalAddress, err = btcutil.NewAddressWitnessScriptHash(
+			btcutil.Hash160(scriptSig),
+			chainParams,
+		)
+	case "zpub", "vpub":
+		// p2wpkh
+		finalAddress, err = btcutil.NewAddressWitnessPubKeyHash(
+			requestedAddress.Hash160()[:],
+			chainParams,
+		)
+	}
+	if err != nil {
+		return "", fmt.Errorf(
+			"failed to derive final address format from extended key: [%s]",
+			err,
+		)
+	}
+
+	return finalAddress.EncodeAddress(), nil
+}
+
+// ConstructUnsignedTransaction produces an unsigned transaction
+func ConstructUnsignedTransaction(
+	previousOutputTransactionHashHex string,
+	previousOutputIndex uint32,
+	previousOutputValue int64,
+	feePerVbyte int64,
+	recipientAddresses []string,
+	chainParams *chaincfg.Params,
+) (*wire.MsgTx, error) {
+	// If the previous output transaction hash is passed as a []byte, can use
+	// chainhash.NewHash.
+	previousOutputTransactionHash, err := chainhash.NewHashFromStr(
+		previousOutputTransactionHashHex,
+	)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"error extracting outpoint transaction hash: [%s]",
+			previousOutputTransactionHash,
+		)
+	}
+
+	// The witness signature field is the DER signature followed by the hash type.
+	// We write a dummy signature with 73 0 bytes. DER signatures vary in encoding
+	// between 71, 72, and 73 bytes, so we choose the longest for fee purposes.
+	dummySignatureForWitness := []byte{
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0,
+		0, // one more dummy byte for the SigHashType
+	}
+	// The compressed public key requires 33 bytes.
+	dummyCompressedPublicKeyForWitness := []byte{
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+		0, 0, 0,
+	}
```

Comment:
Would using `bytes.Repeat([]byte{0}, 74)` be cleaner?

```suggestion
	dummySignatureForWitness := bytes.Repeat([]byte{0}, 74) // one more dummy byte for the SigHashType
	
	// The compressed public key requires 33 bytes.
	dummyCompressedPublicKeyForWitness := bytes.Repeat([]byte{0}, 33)
```

Reply:
I like it way more :100: